### PR TITLE
CHIA-1247 Use blockchain_constants to ensure hard fork behavior in test_fill_rate_block_validation

### PR DESCRIPTION
**Pull request by AmineKhaldi**
### Purpose:

The current logic for validating a mempool transaction generates a block and runs it, and that behavior differs between pre and post hard fork (`run_block_generator` vs `run_block_generator2` respectively).

This ensures that we're testing the hard fork behavior in `test_fill_rate_block_validation`.

### Current Behavior:

The constants used do not trigger hard fork behavior.

### New Behavior:

The new constants trigger the hard fork behavior.

### CODE REVIEW

File: chia/_tests/core/mempool/test_mempool_manager.py [hunk 1]
```diff
@@ -1955,14 +1954,14 @@ async def get_coin_records(coin_ids: Collection[bytes32]) -> List[CoinRecord]:
...
         # because of the spend bundle aggregation that creates the block bundle.
-        (TEST_FILL_RATE_ITEM_COST * 2, 2, TEST_FILL_RATE_ITEM_COST * 2 - 61_829),
+        (TEST_FILL_RATE_ITEM_COST * 2, 2, TEST_FILL_RATE_ITEM_COST * 2 - 156_020),
         # Here we set the block cost limit to twice the test items' cost - 1,
         # so we expect only one of the two test items to get included in the block.
         (TEST_FILL_RATE_ITEM_COST * 2 - 1, 1, TEST_FILL_RATE_ITEM_COST),
     ],
 )
 async def test_fill_rate_block_validation(
-    consensus_mode: ConsensusMode,
```
Comment (arvidn) [inline chia/_tests/core/mempool/test_mempool_manager.py:1965]: apparently we did not get a warning on this unused parameter. is this still exempt from mypy?

File: chia/_tests/core/mempool/test_mempool_manager.py [hunk 2]
```diff
@@ -1955,14 +1954,14 @@ async def get_coin_records(coin_ids: Collection[bytes32]) -> List[CoinRecord]:
         # expect both test items to get included in the block.
         # NOTE: The expected block cost is smaller than the sum of items' costs
         # because of the spend bundle aggregation that creates the block bundle.
-        (TEST_FILL_RATE_ITEM_COST * 2, 2, TEST_FILL_RATE_ITEM_COST * 2 - 61_829),
+        (TEST_FILL_RATE_ITEM_COST * 2, 2, TEST_FILL_RATE_ITEM_COST * 2 - 156_020),
```
Comment (arvidn) [inline chia/_tests/core/mempool/test_mempool_manager.py:1957]: this difference increasing suggests that the "effect of compression" was not avoided, doesn't it?
Do you understand why this changes?

File: chia/_tests/core/mempool/test_mempool_manager.py [hunk 3]
```diff
@@ -1985,21 +1984,25 @@ async def send_to_mempool(full_node: FullNodeSimulator, spend_bundle: SpendBundl
...
-        puzzle = SerializedProgram.to((1, [[ConditionOpcode.REMARK, b"1" * 12_000]]))
-        ph = puzzle.get_tree_hash()
-        for _ in range(3):
-            await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(ph))
-        coin_records = await full_node_api.full_node.coin_store.get_coin_records_by_puzzle_hash(False, ph)
-        coins = [cr.coin for cr in coin_records if cr.coin.amount == 250_000_000_000]
+        coins_and_puzzles = []
+        # Create different puzzles to avoid the effects of block compression
+        for i in (1, 2):
+            puzzle = SerializedProgram.to((1, [[ConditionOpcode.REMARK, bytes(str(i) * 12_000, "utf-8")]]))
```
Comment (arvidn) [inline chia/_tests/core/mempool/test_mempool_manager.py:1990]:
```suggestion
            puzzle = SerializedProgram.to((1, [[ConditionOpcode.REMARK, bytes([i] * 12_000]]))
```
it seems simpler to build a bytes buffer directly

**Review by arvidn — COMMENTED** [review]
I was under the impression that our farmer code alway use compression, regardless of constants. Can you describe how the constants affect the block being farmed?

File: chia/_tests/core/mempool/test_mempool_manager.py [hunk 4]
```diff
@@ -1943,7 +1942,7 @@ async def get_coin_records(coin_ids: Collection[bytes32]) -> List[CoinRecord]:
         assert e.code == expected
 
 
-TEST_FILL_RATE_ITEM_COST = 144_785_330
+TEST_FILL_RATE_ITEM_COST = 144_744_040
```
Comment (arvidn) [inline chia/_tests/core/mempool/test_mempool_manager.py:1945]: the fact that this decreased also suggests that you had some "compression effects", doesn't it? Why else did it drop?
If you spend two coins that have the same parent coin ID, that 32 byte field will be compressed, even if the puzzle reveals are different

**Comment by arvidn** [comment]
I think we should hold off on this until after we've transitioned to rust TX validation.

File: chia/_tests/core/mempool/test_mempool_manager.py (hunk 4 above)
Comment (AmineKhaldi) [reply chia/_tests/core/mempool/test_mempool_manager.py:1945]: It's documented that we use a different (parent) coin and a different puzzle, but we cannot completely eliminate the effect of compression, so we still have a small remaining effect.

File: chia/_tests/core/mempool/test_mempool_manager.py [hunk 5]
```diff
@@ -1985,21 +1986,26 @@ async def send_to_mempool(full_node: FullNodeSimulator, spend_bundle: SpendBundl
...
-        for coin in coins:
+        for coin, puzzle in coins_and_puzzles:
             coin_spend = make_spend(coin, puzzle, SerializedProgram.to([]))
             sb = SpendBundle([coin_spend], G2Element())
             await send_to_mempool(full_node_api, sb)
             sbs_info.append((coin.name(), puzzle, sb.name()))
         return sbs_info
 
-    constants = test_constants.replace(MAX_BLOCK_COST_CLVM=max_block_clvm_cost)
+    constants = blockchain_constants.replace(MAX_BLOCK_COST_CLVM=max_block_clvm_cost)
```
Comment (arvidn) [inline chia/_tests/core/mempool/test_mempool_manager.py:2008]: I take it this is the key change. previously we always tested the validation as if we were producing blocks for the pre-hard fork chain. is that right?

Comment (AmineKhaldi) [reply chia/_tests/core/mempool/test_mempool_manager.py:2008]: Indeed.

**Comment by arvidn** [comment]
my understanding is that the fundamental fix is to actually use the constants fixture in the test. Right now, we run the test for every consensus mode, but we use the same constants every time.

With this change, spend bundles will be compressed when their cost is calculated and therefore changed.

**Comment by arvidn** [comment]
my understanding is that the fundamental fix is to actually use the constants fixture in the test. Right now, we run the test for every consensus mode, but we use the same constants every time.

With this change, spend bundles will be compressed when their cost is calculated and therefore changed.

**Review by arvidn — APPROVED** [review]
